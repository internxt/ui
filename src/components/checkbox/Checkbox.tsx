import * as RadixCheckbox from '@radix-ui/react-checkbox';
import { CheckIcon, DividerHorizontalIcon } from '@radix-ui/react-icons';
import { useState } from 'react';

const Checkbox = ({
  asChild = false,
  defaultChecked = true,
  checked = true,
  disabled = false,
  required,
  name,
  value,
}: {
  asChild?: boolean;
  defaultChecked?: boolean | 'indeterminate';
  checked?: boolean | 'indeterminate';
  disabled?: boolean;
  required?: boolean;
  name?: string;
  value?: string;
}): JSX.Element => {
  const [isChecked, setIsChecked] = useState<boolean | 'indeterminate'>(checked);
  const [isDefaultChecked, setIsDefaultChecked] = useState<boolean | 'indeterminate'>(defaultChecked);
  const isIndetermiante = isChecked === 'indeterminate' || isDefaultChecked === 'indeterminate';

  const onCheckedChange = () => {
    if (isIndetermiante) {
      setIsChecked(false);
      setIsDefaultChecked(false);
    } else {
      setIsChecked(!isChecked);
      setIsDefaultChecked(!isChecked);
    }
  };

  return (
    <RadixCheckbox.Root
      className={`h-5 w-5 rounded flex justify-center items-center 
      ${!disabled && isChecked && 'bg-primary hover:bg-primary-dark'}
      ${disabled && !isChecked && 'bg-surface border border-gray-10'}
      ${!disabled && !isChecked && 'bg-surface border border-gray-30'}
      ${disabled && isChecked && 'bg-gray-20'}
      `}
      asChild={asChild}
      defaultChecked={isDefaultChecked}
      checked={isChecked}
      onCheckedChange={onCheckedChange}
      disabled={disabled}
      required={required}
      name={name}
      value={value}
    >
      <RadixCheckbox.Indicator className="">
        {isChecked === 'indeterminate' && <DividerHorizontalIcon className="text-white" />}
        {isChecked === true && <CheckIcon className="text-white" />}
      </RadixCheckbox.Indicator>
    </RadixCheckbox.Root>
  );
};

export default Checkbox;
