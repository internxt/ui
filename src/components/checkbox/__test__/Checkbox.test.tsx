import React from 'react';
import { describe, expect, it } from 'vitest';
import Checkbox from '../Checkbox';
import renderer from 'react-test-renderer';

describe('Button component', () => {
  it('Checkbox disabled and not checked should render correctly', () => {
    const button = renderer.create(<Checkbox disabled={true} checked={false} />).toJSON();
    expect(button).toMatchSnapshot();
  });

  it('Checkbox enabled and not checked should render correctly', () => {
    const button = renderer.create(<Checkbox checked={false} />).toJSON();
    expect(button).toMatchSnapshot();
  });

  it('Checkbox checked should render correctly', () => {
    const button = renderer.create(<Checkbox checked={true} />).toJSON();
    expect(button).toMatchSnapshot();
  });

  it('Checkbox disabled and checked should render correctly', () => {
    const button = renderer.create(<Checkbox disabled={true} checked={true} />).toJSON();
    expect(button).toMatchSnapshot();
  });

  it('Checkbox enabled and indetermiante should render correctly', () => {
    const button = renderer.create(<Checkbox checked="indeterminate" />).toJSON();
    expect(button).toMatchSnapshot();
  });

  it('Checkbox disabled and indetermiante should render correctly', () => {
    const button = renderer.create(<Checkbox disabled={true} checked="indeterminate" />).toJSON();
    expect(button).toMatchSnapshot();
  });
});
