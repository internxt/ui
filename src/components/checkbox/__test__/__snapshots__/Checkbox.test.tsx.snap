// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button component > Checkbox checked should render correctly 1`] = `
[
  <button
    aria-checked={true}
    className="h-5 w-5 rounded flex justify-center items-center 
      bg-primary hover:bg-primary-dark
      false
      false
      false
      "
    data-state="checked"
    disabled={false}
    onClick={[Function]}
    onKeyDown={[Function]}
    role="checkbox"
    type="button"
    value="on"
  >
    <span
      className=""
      data-state="checked"
      style={
        {
          "pointerEvents": "none",
        }
      }
    >
      <svg
        className="text-white"
        fill="none"
        height="15"
        viewBox="0 0 15 15"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clipRule="evenodd"
          d="M11.4669 3.72684C11.7558 3.91574 11.8369 4.30308 11.648 4.59198L7.39799 11.092C7.29783 11.2452 7.13556 11.3467 6.95402 11.3699C6.77247 11.3931 6.58989 11.3355 6.45446 11.2124L3.70446 8.71241C3.44905 8.48022 3.43023 8.08494 3.66242 7.82953C3.89461 7.57412 4.28989 7.55529 4.5453 7.78749L6.75292 9.79441L10.6018 3.90792C10.7907 3.61902 11.178 3.53795 11.4669 3.72684Z"
          fill="currentColor"
          fillRule="evenodd"
        />
      </svg>
    </span>
  </button>,
  <input
    aria-hidden={true}
    defaultChecked={true}
    disabled={false}
    style={
      {
        "margin": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "transform": "translateX(-100%)",
      }
    }
    tabIndex={-1}
    type="checkbox"
    value="on"
  />,
]
`;

exports[`Button component > Checkbox disabled and checked should render correctly 1`] = `
[
  <button
    aria-checked={true}
    className="h-5 w-5 rounded flex justify-center items-center 
      false
      false
      false
      bg-gray-20
      "
    data-disabled=""
    data-state="checked"
    disabled={true}
    onClick={[Function]}
    onKeyDown={[Function]}
    role="checkbox"
    type="button"
    value="on"
  >
    <span
      className=""
      data-disabled=""
      data-state="checked"
      style={
        {
          "pointerEvents": "none",
        }
      }
    >
      <svg
        className="text-white"
        fill="none"
        height="15"
        viewBox="0 0 15 15"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clipRule="evenodd"
          d="M11.4669 3.72684C11.7558 3.91574 11.8369 4.30308 11.648 4.59198L7.39799 11.092C7.29783 11.2452 7.13556 11.3467 6.95402 11.3699C6.77247 11.3931 6.58989 11.3355 6.45446 11.2124L3.70446 8.71241C3.44905 8.48022 3.43023 8.08494 3.66242 7.82953C3.89461 7.57412 4.28989 7.55529 4.5453 7.78749L6.75292 9.79441L10.6018 3.90792C10.7907 3.61902 11.178 3.53795 11.4669 3.72684Z"
          fill="currentColor"
          fillRule="evenodd"
        />
      </svg>
    </span>
  </button>,
  <input
    aria-hidden={true}
    defaultChecked={true}
    disabled={true}
    style={
      {
        "margin": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "transform": "translateX(-100%)",
      }
    }
    tabIndex={-1}
    type="checkbox"
    value="on"
  />,
]
`;

exports[`Button component > Checkbox disabled and indetermiante should render correctly 1`] = `
[
  <button
    aria-checked="mixed"
    className="h-5 w-5 rounded flex justify-center items-center 
      false
      false
      false
      bg-gray-20
      "
    data-disabled=""
    data-state="indeterminate"
    disabled={true}
    onClick={[Function]}
    onKeyDown={[Function]}
    role="checkbox"
    type="button"
    value="on"
  >
    <span
      className=""
      data-disabled=""
      data-state="indeterminate"
      style={
        {
          "pointerEvents": "none",
        }
      }
    >
      <svg
        className="text-white"
        fill="none"
        height="15"
        viewBox="0 0 15 15"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clipRule="evenodd"
          d="M2 7.5C2 7.22386 2.22386 7 2.5 7H12.5C12.7761 7 13 7.22386 13 7.5C13 7.77614 12.7761 8 12.5 8H2.5C2.22386 8 2 7.77614 2 7.5Z"
          fill="currentColor"
          fillRule="evenodd"
        />
      </svg>
    </span>
  </button>,
  <input
    aria-hidden={true}
    defaultChecked={false}
    disabled={true}
    style={
      {
        "margin": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "transform": "translateX(-100%)",
      }
    }
    tabIndex={-1}
    type="checkbox"
    value="on"
  />,
]
`;

exports[`Button component > Checkbox disabled and not checked should render correctly 1`] = `
[
  <button
    aria-checked={false}
    className="h-5 w-5 rounded flex justify-center items-center 
      false
      bg-surface border border-gray-10
      false
      false
      "
    data-disabled=""
    data-state="unchecked"
    disabled={true}
    onClick={[Function]}
    onKeyDown={[Function]}
    role="checkbox"
    type="button"
    value="on"
  />,
  <input
    aria-hidden={true}
    defaultChecked={false}
    disabled={true}
    style={
      {
        "margin": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "transform": "translateX(-100%)",
      }
    }
    tabIndex={-1}
    type="checkbox"
    value="on"
  />,
]
`;

exports[`Button component > Checkbox enabled and indetermiante should render correctly 1`] = `
[
  <button
    aria-checked="mixed"
    className="h-5 w-5 rounded flex justify-center items-center 
      bg-primary hover:bg-primary-dark
      false
      false
      false
      "
    data-state="indeterminate"
    disabled={false}
    onClick={[Function]}
    onKeyDown={[Function]}
    role="checkbox"
    type="button"
    value="on"
  >
    <span
      className=""
      data-state="indeterminate"
      style={
        {
          "pointerEvents": "none",
        }
      }
    >
      <svg
        className="text-white"
        fill="none"
        height="15"
        viewBox="0 0 15 15"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clipRule="evenodd"
          d="M2 7.5C2 7.22386 2.22386 7 2.5 7H12.5C12.7761 7 13 7.22386 13 7.5C13 7.77614 12.7761 8 12.5 8H2.5C2.22386 8 2 7.77614 2 7.5Z"
          fill="currentColor"
          fillRule="evenodd"
        />
      </svg>
    </span>
  </button>,
  <input
    aria-hidden={true}
    defaultChecked={false}
    disabled={false}
    style={
      {
        "margin": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "transform": "translateX(-100%)",
      }
    }
    tabIndex={-1}
    type="checkbox"
    value="on"
  />,
]
`;

exports[`Button component > Checkbox enabled and not checked should render correctly 1`] = `
[
  <button
    aria-checked={false}
    className="h-5 w-5 rounded flex justify-center items-center 
      false
      false
      bg-surface border border-gray-30
      false
      "
    data-state="unchecked"
    disabled={false}
    onClick={[Function]}
    onKeyDown={[Function]}
    role="checkbox"
    type="button"
    value="on"
  />,
  <input
    aria-hidden={true}
    defaultChecked={false}
    disabled={false}
    style={
      {
        "margin": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "transform": "translateX(-100%)",
      }
    }
    tabIndex={-1}
    type="checkbox"
    value="on"
  />,
]
`;

exports[`Button component > Checkbox indetermiante should render correctly 1`] = `
[
  <button
    aria-checked="mixed"
    className="h-5 w-5 rounded flex justify-center items-center 
      bg-primary hover:bg-primary-dark
      false
      false
      false
      "
    data-state="indeterminate"
    disabled={false}
    onClick={[Function]}
    onKeyDown={[Function]}
    role="checkbox"
    type="button"
    value="on"
  >
    <span
      className=""
      data-state="indeterminate"
      style={
        {
          "pointerEvents": "none",
        }
      }
    >
      <svg
        className="text-white"
        fill="none"
        height="15"
        viewBox="0 0 15 15"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clipRule="evenodd"
          d="M2 7.5C2 7.22386 2.22386 7 2.5 7H12.5C12.7761 7 13 7.22386 13 7.5C13 7.77614 12.7761 8 12.5 8H2.5C2.22386 8 2 7.77614 2 7.5Z"
          fill="currentColor"
          fillRule="evenodd"
        />
      </svg>
    </span>
  </button>,
  <input
    aria-hidden={true}
    defaultChecked={false}
    disabled={false}
    style={
      {
        "margin": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "transform": "translateX(-100%)",
      }
    }
    tabIndex={-1}
    type="checkbox"
    value="on"
  />,
]
`;

exports[`Button component > Primary button should render correctly 1`] = `
[
  <button
    aria-checked={true}
    className="h-5 w-5 rounded flex justify-center items-center 
      bg-primary hover:bg-primary-dark
      false
      false
      false
      "
    data-state="checked"
    disabled={false}
    onClick={[Function]}
    onKeyDown={[Function]}
    role="checkbox"
    type="button"
    value="on"
  >
    <span
      className=""
      data-state="checked"
      style={
        {
          "pointerEvents": "none",
        }
      }
    >
      <svg
        className="text-white"
        fill="none"
        height="15"
        viewBox="0 0 15 15"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clipRule="evenodd"
          d="M11.4669 3.72684C11.7558 3.91574 11.8369 4.30308 11.648 4.59198L7.39799 11.092C7.29783 11.2452 7.13556 11.3467 6.95402 11.3699C6.77247 11.3931 6.58989 11.3355 6.45446 11.2124L3.70446 8.71241C3.44905 8.48022 3.43023 8.08494 3.66242 7.82953C3.89461 7.57412 4.28989 7.55529 4.5453 7.78749L6.75292 9.79441L10.6018 3.90792C10.7907 3.61902 11.178 3.53795 11.4669 3.72684Z"
          fill="currentColor"
          fillRule="evenodd"
        />
      </svg>
    </span>
  </button>,
  <input
    aria-hidden={true}
    defaultChecked={true}
    disabled={false}
    style={
      {
        "margin": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "transform": "translateX(-100%)",
      }
    }
    tabIndex={-1}
    type="checkbox"
    value="on"
  />,
]
`;
